joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -85,31 +89,147 @@ JointTrajectoryController::update()
     return CONTROLLER_INTERFACE_RET_SUCCESS;
   }
 
-  // when we receive a traj msg
-  if (traj_point_active_ptr_ && (*traj_point_active_ptr_)->is_empty() == false) {
-    // find next new point for current timestamp
-    auto traj_point_ptr = (*traj_point_active_ptr_)->sample(lifecycle_node_->now());
-    // find next new point for current timestamp
-    // set cmd only if a point is found
-    if (traj_point_ptr == (*traj_point_active_ptr_)->end()) {
-      return CONTROLLER_INTERFACE_RET_SUCCESS;
-    }
+  auto resize_joint_trajectory_point =
+    [](trajectory_msgs::msg::JointTrajectoryPoint & point, int size)
+    {
+      point.positions.resize(size);
+      point.velocities.resize(size);
+      point.accelerations.resize(size);
+    };
+
+  JointTrajectoryPoint state_current, state_desired, state_error;
+  size_t joint_num = registered_joint_state_handles_.size();
+  resize_joint_trajectory_point(state_current, joint_num);
 
-    // check if new point ptr points to the same as previous point
-    if (prev_traj_point_ptr_ == traj_point_ptr) {
-      return CONTROLLER_INTERFACE_RET_SUCCESS;
+  // current state update
+  for (uint index = 0; index < joint_num; ++index) {
+    auto & joint_state = registered_joint_state_handles_[index];
+    state_current.positions[index] = joint_state->get_position();
+    state_current.velocities[index] = joint_state->get_velocity();
+    state_current.accelerations[index] = 0.0;
+  }
+  state_current.time_from_start.set__sec(0);
+
+  std::lock_guard<std::mutex> guard(trajectory_mtx_);
+
+  // currently carrying out a trajectory
+  if (traj_point_active_ptr_ && (*traj_point_active_ptr_)->has_traj_msg() == false) {
+    // if sampling the first time, set the point before you sample
+    if (!(*traj_point_active_ptr_)->is_sampled_already()) {
+      (*traj_point_active_ptr_)->set_point_before_traj(lifecycle_node_->now(), state_current);
     }
+    resize_joint_trajectory_point(state_error, joint_num);
+
+    // find segment for current timestamp
+    TrajectoryPointConstIter start_segment_itr, end_segment_itr;
+    bool valid_point = (*traj_point_active_ptr_)->sample(
+      lifecycle_node_->now(), state_desired,
+      start_segment_itr, end_segment_itr);
+
+    auto compute_error_for_joint = [](JointTrajectoryPoint & error, int index,
+        const JointTrajectoryPoint & current, const JointTrajectoryPoint & desired)
+      {
+        // error defined as the difference between current and desired
+        error.positions[index] = angles::shortest_angular_distance(
+          current.positions[index], desired.positions[index]);
+
+        if (!desired.velocities.empty()) {
+          error.velocities[index] = current.velocities[index] - desired.velocities[index];
+        } else {
+          error.velocities[index] = 0.0;
+        }
+        error.accelerations[index] = 0.0;
+      };
```

Comment:
Consider moving the two lambdas together for readability.

Reply:
fixed aec9f2d